src/v/io/persistence.cc

```diff
@@ -0,0 +1,327 @@
+          len,
+          memory_dma_alignment(),
+          disk_read_dma_alignment());
+        err.has_value()) {
+        return std::move(err.value());
+    }
+    return read(pos, buf, len);
+}
+
+seastar::future<size_t> memory_persistence::memory_file::dma_write(
+  uint64_t pos, const char* buf, size_t len) noexcept {
+    if (auto err = check_alignment(
+          "dma_write",
+          pos,
+          buf,
+          len,
+          memory_dma_alignment(),
+          disk_write_dma_alignment());
+        err.has_value()) {
+        return std::move(err.value());
+    }
+    return write(pos, buf, len).then([this, pos](size_t written) {
+        size_ = std::max(size_, pos + written);
+        return written;
+    });
+}
+
+seastar::future<> memory_persistence::memory_file::close() noexcept {
+    co_return;
+}
+
+uint64_t
+memory_persistence::memory_file::disk_read_dma_alignment() const noexcept {
+    return persistence_->disk_read_dma_alignment_;
+}
+
+uint64_t
+memory_persistence::memory_file::disk_write_dma_alignment() const noexcept {
+    return persistence_->disk_write_dma_alignment_;
+}
+
+uint64_t
+memory_persistence::memory_file::memory_dma_alignment() const noexcept {
+    return persistence_->memory_dma_alignment_;
+}
+
+seastar::future<>
+memory_persistence::memory_file::ensure_capacity(size_t size) {
+    // futurized to avoid reactor stalls for large capacity expansions
+    return seastar::do_until(
+      [this, size] { return capacity_ >= size; },
+      [this] {
+          data_.emplace(capacity_, chunk_size);
+          capacity_ += chunk_size;
+          return seastar::make_ready_future<>();
+      });
+}
+
+seastar::future<size_t>
+memory_persistence::memory_file::read(uint64_t pos, char* buffer, size_t len) {
+    if (pos >= size_) {
+        co_return 0;
+    }
+
+    std::span output(buffer, len);
+
+    auto it = data_.upper_bound(pos);
+    // first chunk that starts strictly after pos ensures that begin() will
+    // never be returned since zero is the smallest possible offset.
+    it = std::prev(it);
+
+    size_t bytes_read = 0;
+    len = std::min(len, size_ - pos);
+    while (len) {
+        assert(it != data_.end());
+
+        auto chunk_pos = pos - it->first;
+        auto chunk_len = std::min(len, chunk_size - chunk_pos);
+
+        std::span chunk(it->second.get(), it->second.size());
+        auto src = chunk.subspan(chunk_pos);
+        auto dst = output.subspan(bytes_read);
+        std::copy_n(src.begin(), chunk_len, dst.begin());
+
+        bytes_read += chunk_len;
+        pos += chunk_len;
+        len -= chunk_len;
+
+        ++it;
+
+        co_await seastar::maybe_yield();
+    }
+
+    co_return bytes_read;
+}
+
+seastar::future<size_t> memory_persistence::memory_file::write(
+  uint64_t pos, const char* buf, size_t len) {
+    if (len == 0) {
+        // don't expand capacity with a zero-length write at @pos
+        co_return 0;
+    }
+    co_await ensure_capacity(pos + len);
```

Comment:
My inexperience with disks + dma is showing: is this really how `dma_write` works, you can write to an arbitrary offset and the file is expanded?

Reply:
yeh, should be. write at pos=file_size to expand. there are well defined semantics about what is read when holes are created, but redpanda doesn't write sparse files. i think a good extension here would be to do something like fill holes with some poison data to catch issues. i've added this to the todo list.